### ISSUE TITLE
Should all data classes in our TP be immutable?

### ISSUE THREAD

**Issue by honganhcs** [nus-cs2103-AY2122S2/forum#193]
I noticed that the `Person` class and the classes representing its fields all have their instance members declared as final. Thus, for `EditCommand`, the `execute` method creates a new `Person` object, which replaces the person-to-edit object in the `UniquePersonList`. 
What are the benefits of doing this, compared to modifying the fields in the person-to-edit object?

**Comment by takufunkai** [comment]
Imo, this makes the program (and its classes) a lot simpler to reason about:
- No need for setters
  - Rely on our method of construction to build the object correctly
- No need for extra validation
  - With setters, we need to run validation checks for each of the possible update methods

To some extent I think this is DRY and reduces code complexity

**Comment by YiHern-Lee** [comment]
I agree with @takufunkai.
In the future when the program grows and there is a need to share references to the same object, allowing the mutation of objects may result in unwanted side-effects and make debugging more difficult.

**Comment by TypeDefinition** [comment]
I disagree that `Person` should be immutable. Certain granular things makes sense to be immutable, such as `Phone`, `Address`, `Name` etc.

However, the entire core of our program revolves around modifying a person. In the real world, if a person changed their phone number, they'd replace my entire phone number. But we wouldn't replace the entire person with someone else entirely.

I find that by needlessly making everything immutable, we end up with silly classes like `EditPersonDescriptor`. Combined with the fact every field in `Person` is so highly coupled, we end up with a problem where simply adding a new field like `Remark` requires changes to an unreasonably large number of files, and is frankly, unmaintainable and much more bug prone than to have a simple, mutable design from the start.

**TLDR: Immutability should be applied when it fits the situation. But not when it conflicts with the core purpose of the application.**

**Comment by chuabingquan** [comment]
I do think there's a case to be made for and against writing your code in a completely immutable manner as proposed by @takufunkai and @YiHern-Lee, but I don't think the case against it as described by @TypeDefinition holds.

While there are advantages to modelling our programs as closely as possible to represent real world entities and their behaviours, we mustn't forget that at the end of the day, such programs are just simplistic abstractions of these entities and their behaviours.

As such, it shouldn't matter that updating the `name` of a `Person` would require us to replace it with a new `Person` instance, the same way it shouldn't matter that modifying the `name` field of a `Person` doesn't require us to model and implement the various legal frameworks and processes of different countries just so we can facilitate a name change in our program. In other words, while programs can attempt to model the real world, it doesn't need to (and it isn't realistic to) do so to the last atom.

Furthermore, there exist languages like [Rust](https://www.rust-lang.org/), [Haskell](https://www.haskell.org/), [OCaml](https://ocaml.org/), and more, that supports immutability by default, along with a certain flavour of [variable shadowing](https://en.wikipedia.org/wiki/Variable_shadowing). As such, programs like the following example here are commonplace and idiomatic as well:
```hs
-- Updating a Person's age in Haskell:
let john = Person { name = "John", age = 25 }
-- The operation on the right produces a new instance of `john` with his updated age.
let john = john { age = 21 }
{- Since `john` in the second line shadows the `john` of the first line, 
future references to `john` would refer to the instance on the second line. -}
... 
```
With respect to the example on `EditPersonDescriptor`, I'm unable to see how Java programs that are written in an immutable fashion would lead to such unnecessary boilerplate. If anything, this strikes me as a poorly formulated design, one reminiscent of [jokes about over-abstraction in enterprise Java programs](https://i.redd.it/rqsz6jzvfif11.png). In fact, Java programs that are written in an immutable fashion would largely resemble their mutable counterparts, albeit with more re-assignment or return statements. As an example, let's compare the difference between two hypothetical `Person` implementations, one that's mutable, and one that's not:

**Mutable Implementation**
```java
...
private final Name name;
...
public void setName(String newName) {
    this.name.update(newName);
}
...
johnny.setName("John");
```

**Immutable Implementation**
```java
...
private Name name;
...
public Person setName(String newName) {
    // You can probably abstract this further to make the program more neater;
    // e.g., return this.cloneWith(newName); -- this method can be overloaded.
    return new Person(..., this.name.update(newName));
}
...
final Person john = johnny.setName("John");
```

As with the example above, there's no need for a class like `EditPersonDescriptor`. In exchange, your immutable implementation is now much loosely coupled, making it easier to unit-test and reason about your program. Arguably, if you do end up needing an `EditPersonDescriptor`, it's much more likely a sign that your `Person` class is too bloated.

In general, I concur with the rationale for immutability that's put forth by @takufunkai and @YiHern-Lee. However, Java programs that are written in an immutable fashion are likely to be unidiomatic and anti-pattern to well established Java conventions, which often utilises mutation. Consequently, other Java programmers might find it hard to takeover or maintain your codebase. In addition, if good performance is critical for your program, a language like Java that's not designed with immutability in mind could suffer with respect to high memory usage. In other words, you would have to weigh these pros and cons when deciding on whether to write your Java programs in an immutable manner.

**Comment by honganhcs** [comment]
Thank you for all your inputs! I realised that one reason why editing a person requires replacing the Person object in the UniquePersonList is so that listeners to the ObservableList<Person> will know that there is an update to the list. I think that listeners to ObservableLists only listen to changes in the list and not to changes in an item in the list. May I know if my understanding is correct Prof @damithc?

**Comment by damithc** [comment]
> May I know if my understanding is correct Prof @damithc?

@honganhcs most likely so. I can't confirm as I wasn't heavily involved in writing the code.

**Comment by damithc** [comment]
Glad to see this kind of discussions on design decisions. Do keep it up as this is an interesting topic.

In general, we encourage you to question design decisions in the current code and explore alternatives. The work done when exploring alternatives can earn credit (as they require effort) and also can be submitted for [CS2103R](https://nus-cs2103-ay2122s2.github.io/website/admin/appendixG-CS2103R.html).

Things to keep in mind:
1. Overuse or blind application of principles and best practices is not good. Everything has pros and cons, and some have trade-offs/conflicts with each other. But they are not totally useless either (i.e., don't be hasty when discarding them)
2. Keep discussions civil and professional. Don't be too dogmatic or condescending of others' views. 
3. Keep an open mind when critiquing existing code. Often there is more to it than meets the eye at first. Some past decisions may  have been misguided to begin with, or justified at the time but no longer so due to subsequent evolution of the code (e.g., AB3 is scaled back version of AB4 which was going for an AB5 at the time), or could still be valid despite the first impressions due to other factors not immediately obvious.
4. Often, a concrete implementation of an alternative is the best way to find the full implications of it (pros/cons), compared to a hypothetical argument with small code examples i.e., the proof is in the pudding.

In this particular case, I remember the `Person` class used to be mutable at first and there was a `ReadOnlyPerson` interface that was used when `Person` objects were passed to a client that shouldn't be updating the objects. At some point, that design was updated to make the `Person` class immutable instead. It's pity that the rationale for that was not documented, which also goes to show the value of good developer documentation.
